Dalamud/Game/Gui/Nameplates/NameplateGui.cs

```diff
@@ -0,0 +1,211 @@
+using System.Collections.Generic;
+using System.Runtime.InteropServices;
+
+using Dalamud.Game.Addon.Lifecycle;
+using Dalamud.Game.Addon.Lifecycle.AddonArgTypes;
+using Dalamud.Game.ClientState.Objects;
+using Dalamud.Game.ClientState.Objects.Types;
+using Dalamud.Game.Gui.Nameplates.Model;
+using Dalamud.Game.Text.SeStringHandling;
+using Dalamud.Hooking;
+using Dalamud.IoC.Internal;
+using Dalamud.Logging.Internal;
+using Dalamud.Memory;
+using Dalamud.Utility.Signatures;
+
+using FFXIVClientStructs.FFXIV.Client.System.String;
+using FFXIVClientStructs.FFXIV.Client.UI;
+using FFXIVClientStructs.FFXIV.Component.GUI;
+
+using TerraFX.Interop.Windows;
+
+namespace Dalamud.Game.Gui.Nameplates;
+
+/// <summary>
+/// This class handles interacting with native Nameplate update events and management.
+/// </summary>
+[ServiceManager.EarlyLoadedService]
+internal class NameplateGui : IInternalDisposableService, INameplatesGui
+{
+    [ServiceManager.ServiceDependency]
+    private readonly GameGui gameGui = Service<GameGui>.Get();
+
+    [ServiceManager.ServiceDependency]
+    private readonly ObjectTable objectTable = Service<ObjectTable>.Get();
+
+    [ServiceManager.ServiceDependency]
+    private readonly AddonLifecycle addonLifecycle = Service<AddonLifecycle>.Get();
+
+    [Signature("E8 ?? ?? ?? ?? E9 ?? ?? ?? ?? 48 8B 5C 24 ?? 45 38 BE", DetourName = nameof(HandleSetPlayerNameplateDetour))]
+    private readonly Hook<SetPlayerNameplateDetourDelegate> setPlayerNameplateDetourHook;
+    private nint namePlatePtr;
+
+    /// <summary>
+    /// Initializes a new instance of the <see cref="NameplateGui"/> class.
+    /// </summary>
+    [ServiceManager.ServiceConstructor]
+    private NameplateGui(TargetSigScanner sigScanner)
+    {
+        // Pointers
+        this.addonLifecycle.RegisterListener(new AddonLifecycleEventListener(AddonEvent.PostSetup, "NamePlate", this.AddonLifecycle_Event));
+        this.addonLifecycle.RegisterListener(new AddonLifecycleEventListener(AddonEvent.PreFinalize, "NamePlate", this.AddonLifecycle_Event));
+
+        // Hooks
+        SignatureHelper.Initialize(this);
+        this.setPlayerNameplateDetourHook.Enable();
+    }
+
+    private unsafe delegate nint SetPlayerNameplateDetourDelegate(nint playerNameplateObjectPtr, bool isTitleAboveName, bool isTitleVisible, nint titlePtr, nint namePtr, nint freeCompanyPtr, nint prefix, int iconId);
+
+    /// <inheritdoc/>
+    public event INameplatesGui.OnNameplateUpdateDelegate OnNameplateUpdate;
+
+    /// <inheritdoc/>
+    public void DisposeService()
+    {
+        this.setPlayerNameplateDetourHook.Dispose();
+    }
+
+    /// <inheritdoc/>
+    public T? GetNameplateGameObject<T>(INameplateObject nameplateObject) where T : GameObject
+    {
+        return this.GetNameplateGameObject<T>(nameplateObject.Pointer);
+    }
+
+    /// <inheritdoc/>
+    public T? GetNameplateGameObject<T>(nint nameplateObjectPtr) where T : GameObject
```

Comment:
I wonder if it makes sense to cache the result of this in `NameplateObject` or similar. I think finding either the index or the object is going to be necessary for most users as that's how they're likely to either find the nameplate owner. As-is each consumer is likely to repeat this call, so we'll end up looking up the index and scanning the object table multiple times.

Reply:
Yup, makes sence. I removed the generic type and added a property to `INameplateObject`. Also removed `GetNameplateGameObject<T>` from `INameplateGui`.